docs/Configuration.md
# Configuration

## Debugging

In case a model plugin doesn't work correctly (ios, procurve, etc.), you can enable live debugging of SSH/Telnet sessions. Just add a `debug` option containing the value true to the `input` section. The log files will be created depending on the parent directory of the logfile option.

The following example will log an active ssh/telnet session `/home/oxidized/.config/oxidized/log/<IP-Address>-<PROTOCOL>`. The file will be truncated on each consecutive ssh/telnet session, so you need to put a `tailf` or `tail -f` on that file!

```yaml
log: /home/oxidized/.config/oxidized/log

...

input:
  default: ssh, telnet
  debug: true
  ssh:
    secure: false
  http:
    ssl_verify: true
```

## Privileged mode

To start privileged mode before pulling the configuration, Oxidized needs to send the enable command. You can globally enable this, by adding the following snippet to the global section of the configuration file.

```yaml
vars:
   enable: S3cre7
```

## Removing secrets

To strip out secrets from configurations before storing them, Oxidized needs the `remove_secret` flag. You can globally enable this by adding the following snippet to the global section of the configuration file.

```yaml
vars:
  remove_secret: true
```

Device models that contain substitution filters to remove sensitive data will now be run on any fetched configuration.

As a partial example from ios.rb:

```ruby
  cmd :secret do |cfg|
    cfg.gsub! /^(snmp-server community).*/, '\\1 <configuration removed>'
    (...)
    cfg
  end
```

The above strips out snmp community strings from your saved configs.

**NOTE:** Removing secrets reduces the usefulness as a full configuration backup, but it may make sharing configs easier.

## Disabling SSH exec channels

Oxidized uses exec channels to make information extraction simpler, but there are some situations where this doesn't work well, e.g. configuring devices. This feature can be turned off by setting the `ssh_no_exec`
variable.

```yaml
vars:
  ssh_no_exec: true
```

## Disabling SSH keepalives

Oxidized SSH input makes use of SSH keepalives to prevent timeouts from slower devices and to quickly tear down stale sessions in larger deployments. There have been reports of SSH keepalives breaking compatibility with certain OS types. They can be disabled using the `ssh_no_keepalive` variable on a per-node basis (by specifying it in the source) or configured application-wide.

```yaml
vars:
  ssh_no_keepalive: true
```

## SSH Auth Methods

By default, Oxidized registers the following auth methods: `none`, `publickey` and `password`. However you can configure this globally, by groups, models or nodes.

```yaml
vars:
  auth_methods: [ "none", "publickey", "password", "keyboard-interactive" ]
```

## Public Key Authentication with SSH

Instead of password-based login, Oxidized can make use of key-based SSH authentication.

You can tell Oxidized to use one or more private keys globally, or specify the key to be used on a per-node basis. The latter can be done by mapping the `ssh_keys` variable through the active source.

Global:

```yaml
vars:
  ssh_keys: "~/.ssh/id_rsa"
```

Per-Node:

```yaml
...
map:
  name: 0
  model: 1
vars_map:
  enable: 2
  ssh_keys: 3
...
```

If you are using a non-standard path, especially when copying the private key via a secured channel, make sure that the permissions are set correctly:

```bash
foo@bar:~$ ls -la ~/.ssh/
total 20
drwx------ 2 oxidized oxidized 4096 Mar 13 17:03 .
drwx------ 5 oxidized oxidized 4096 Mar 13 21:40 ..
-r-------- 1 oxidized oxidized  103 Mar 13 17:03 authorized_keys
-rw------- 1 oxidized oxidized  399 Mar 13 17:02 id_ed25519
-rw-r--r-- 1 oxidized oxidized   94 Mar 13 17:02 id_ed25519.pub
```

Finally, multiple private keys can be specified as an array of file paths, such as `["~/.ssh/id_rsa", "~/.ssh/id_another_rsa"]`.

## SSH Proxy Command

Oxidized can `ssh` through a proxy as well. To do so we just need to set `ssh_proxy` variable with the proxy host information and optionally set the `ssh_proxy_port` with the SSH port if it is not listening on port 22.

This can be provided on a per-node basis by mapping the proper fields from your source.

An example for a `csv` input source that maps the 4th field as the `ssh_proxy` value and the 5th field as `ssh_proxy_port`.

```yaml
...
map:
  name: 0
  model: 1
vars_map:
  enable: 2
  ssh_proxy: 3
  ssh_proxy_port: 4
...
```

## SSH enabling legacy algorithms

When connecting to older firmware over SSH, it is sometimes necessary to enable legacy/disabled settings like KexAlgorithms, HostKeyAlgorithms, MAC or the Encryption.

These settings can be provided on a per-node basis by mapping the ssh_kex, ssh_host_key, ssh_hmac and the ssh_encryption fields from you source.

```yaml
...
map:
  name: 0
  model: 1
vars_map:
  enable: 2
  ssh_kex: 3
  ssh_host_key: 4
  ssh_hmac: 5
  ssh_encryption: 6
...
```

## FTP Passive Mode

Oxidized uses ftp passive mode by default. Some devices require passive mode to be disabled. To do so, we can set `input.ftp.passive` to false - this will make use of FTP active mode.

```yaml
input:
  ftp:
    passive: false
```

## Advanced Configuration

Below is an advanced example configuration. You will be able to (optionally) override options per device. The router.db format used is `hostname:model:username:password:enable_password`. Hostname and model will be the only required options, all others override the global configuration sections.

```yaml
---
username: oxidized
password: S3cr3tx
model: junos
interval: 3600 #interval in seconds
log: ~/.config/oxidized/log
debug: false
threads: 30 # maximum number of threads
# use_max_threads:
# false - the number of threads is selected automatically based on the interval option, but not more than the maximum
# true - always use the maximum number of threads
use_max_threads: false
timeout: 20
retries: 3
prompt: !ruby/regexp /^([\w.@-]+[#>]\s?)$/
crash:
  directory: ~/.config/oxidized/crashes
  hostnames: false
vars:
  enable: S3cr3tx
groups: {}
rest: 127.0.0.1:8888
pid: ~/.config/oxidized/oxidized.pid
input:
  default: ssh, telnet
  debug: false
  ssh:
    secure: false
output:
  default: git
  git:
      user: Oxidized
      email: oxidized@example.com
      repo: "~/.config/oxidized/oxidized.git"
source:
  default: csv
  csv:
    file: ~/.config/oxidized/router.db
    delimiter: !ruby/regexp /:/
    map:
      name: 0
      model: 1
      username: 2
      password: 3
    vars_map:
      enable: 4
model_map:
  cisco: ios
  juniper: junos
```

## Advanced Group Configuration

For group specific credentials

```yaml
groups:
  mikrotik:
    username: admin
    password: blank
  ubiquiti:
    username: ubnt
    password: ubnt
```

Model specific variables within groups

```yaml
groups:
  foo:
    models:
      arista:
        vars:
          ssh_keys: "~/.ssh/id_rsa_foo_arista"
      vyatta:
        vars:
          ssh_keys: "~/.ssh/id_rsa_foo_vyatta"
  bar:
    models:
      routeros:
        vars:
          ssh_keys: "~/.ssh/id_rsa_bar_routeros"
      vyatta:
        vars:
          ssh_keys: "~/.ssh/id_rsa_bar_vyatta"
```

For mapping multiple group values to a common name

```yaml
group_map:
  alias1: groupA
  alias2: groupA
  alias3: groupB
  alias4: groupB
  aliasN: groupZ
  ...
```

add group mapping to a source

```yaml
source:
  ...
  <source>:
    ...
    map:
      model: 0
      name: 1
      group: 2
```

For model specific credentials

You can add 'username: nil' if the device only expects a Password at prompt.

```yaml
models:
  junos:
    username: admin
    password: password
  ironware:
    username: admin
    password: password
    vars:
      enable: enablepassword
  apc_aos:
    username: apc
    password: password
  cisco:
    username: nil
    password: pass
```

## RESTful API and Web Interface

The RESTful API and Web Interface is enabled by configuring the `rest:` parameter in the config file.  This parameter can optionally contain a relative URI.

```yaml
# Listen on http://127.0.0.1:8888/
rest: 127.0.0.1:8888
```

```yaml
# Listen on http://10.0.0.1:8000/oxidized/
rest: 10.0.0.1:8000/oxidized
```

## Triggered backups

A node can be moved to head-of-queue via the REST API `GET/POST /node/next/[NODE]`. This can be useful to immediately schedule a fetch of the configuration after some other event such as a syslog message indicating a configuration update on the device.

In the default configuration this node will be processed when the next job worker becomes available, it could take some time if existing backups are in progress. To execute moved jobs immediately a new job can be added automatically:

```yaml
next_adds_job: true
```

This will allow for a more timely fetch of the device configuration.

## Disabling DNS resolution

In some instances it might not be desirable to attempt to resolve names of nodes. One such use case is when nodes are accessed through an SSH proxy, where the remote end resolves the names differently than the host on which Oxidized runs would.

Names can instead be passed verbatim to the input:

```yaml
resolve_dns: false
```


## Statistics

The `stats:` config section can be used to customize the storage mechanism and amount of per-device statistics maintained by Oxidized. By default, the stats gathered during the last 10 polling attempts are maintained in memory (no persistence) for each device.

Statistics can optionally be stored in [Redis](https://redis.io/) to allow for persistence and better scalability. The [redis-rb](https://github.com/redis/redis-rb) gem must also be installed in order to enable this feature.

```yaml
stats:
  # How many detailed stats should be stored per device?
  # This is used when computing items such as the success/failure rate. Defaults to 10.
  history_size: 10

  # Enable statistics storage in Redis by specifiying a redis:// URL here
  # Set to false to use the standalone in-memory implementation.
  redis_url: 'redis://user:pass@127.0.0.1:6379/16

  # Optionally define how many reconnection attempts will be made to Redis before giving up.
  # Default is 10 attempts
  redis_reconnect_attempts: 10
```


## Environment variables

You can use some environment variables to change default root directories values.

* `OXIDIZED_HOME` may be used to set oxidized configuration directory, which defaults to `~/.config/oxidized`
* `OXIDIZED_LOGS` may be used to set oxidzied logs and crash directories root, which default to `~/.config/oxidized`


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/).

## [Unreleased]

### Added

- add ability to store stats in redis for persistence (@gs-kamnas)
- Extend http source configurations to include read_timeout value
- enterasys800 model for enterasys 800-series fe/ge switches (@javichumellamo)
- add ES3526XA-V2 support in EdgeCOS model (@moisseev)
- model for eltex mes-series switches (@glaubway)
- model for zte c300 and c320 olt (@glaubway)
- model for LANCOM (@systeembeheerder)
- model for Aruba CX switches (@jmurphy5)
- model for NEC IX devices (@mikenowak)
- Added docs for Dell/EMC Networking OS10 devices (@davromaniak)
- model for Zyxel 1308 OLTs (@baldoarturo)
- model for Linksys SRW switches (@glance-)
- model for Cambium ePMP radios (@martydingo)

### Changed

- Better manage of the enable mode in edgeswitch.rb (@agabellini)
- Adds paging support to Enterasys B3/C3 (@piterpunk)
- Allows "Username" as username prompt in Brocade ICX-series devices (@piterpunk) 
- Add show-sensitive flag on export command on Mikrotik RouterOS when remove_secret is off (@kedare)
- rubocop dependency now ~> 0.81.0, the last one with ruby 2.3 support
- change pfSense secret scrubbing to handle new format in 2.4.5+
- Dockerfile rebased to phusion/baseimage-docker bionic-1.0.0
- scrub PoE related messages from routeros config output (@pioto)
- support for d-link dgs-1100 series switches in dlink model (@glaubway)
- enterasys model now works with both ro and rw access (@sargon)
- restore including last configuration change IOS if the change is done by real user. Resolves #1921 (@raunz)
- Additional scrubbing for SonicOS v7 devices (@gerard780)
- improved Telnet support for enterasys (@jplitza)
- Include "show version" output for enterasys (@jplitza)
- xmppdiff now also shows diffs with only removed or only added lines (@jplitza)
- xmppdiff now persists its connection to the XMPP server and MUC (@jplitza)
- routeros no longer backups infos on available updates (@jplitza)
- avoid /humidity hardware field in tmos (F5) to be reported (@albsga)
- version information or OPNsense and PFsense models is now included as XML comments (@pv2b)
- netscaler, backup all partitions (@smallsam)
- only runs SSH proxy commands if the ssh_proxy configuration item has been defined (@jameskirsop)
- updated vrp.rb to correctly parse huawei devices
- asa: information about the configuration change time is deleted

### Fixed

- fixed an issue where Oxidized could not pull config from XOS-devices operating in stacked mode (@DarkCatapulter)
- fixed an issue where Oxidized could not pull config from XOS-devices that have not saved their configuration (@DarkCatapulter)
- improved scrubbing of show chassis in ironware model (@michaelpsomiadis)
- fixed snmp secret handling in netgear model (@CirnoT)
- generalise fortiOS to support new versions and VM based products e.g. FortiManager, FortiAnalyzer and FortiMail, which requires keyboard-interactive auth method. Includes trial-and-error cmd list to retrieve most infromative config. Fixes #2227. Moved 'system ha status' to 'config global' section to support clusters with virtual domains.
- filter next periodic save schedule time in xos model output (@sargon)
- Fix when auto-saved is configured on xos switches  (@trappiz)
- fixed ArubaOS-CX enviroment/system inconsistent values #2297 (@raunz)
- Update AirFiber prompt regex (@murrant)
- System time and running time are now stripped from tplink model output (@spike77453)
- &lt;?xml... line is no longer improperly stripped from OPNsense and PFsense backups (@pv2b)
- fixed an issue where Oxidized timeouts in Brocade ICX-series devices (@piterpunk)
- fixed an issue where EOS config was truncated. Fixes #2038 (@jake2184 @fhibler)
- fixed missing output from routeros version command (@mjbnz)
- Updated fastiron enable password prompt regex (@pepperoni-pi)
- Oxidized will now exit on SIGTERM and SIGINT; SIGHUP will trigger a node list reload and reopen logs (@gs-kamnas)
- fixed an issue where OXIDIZED\_HOME did not appropriately set the main user config file location (@gs-kamnas)

## [0.28.0 - 2020-05-18]

### Added

- add VMWare NSX Edge 6.4+ support (@elmobp)
- add Dell EMC Networking OS10 support (@mmisslin)
- add Centec Networks CNOS (Fiberstore S5800/S5850) support via cnos model (@freddy36)
- include transceiver information in EdgeCOS model (@freddy36)
- add Telco Systems T-Marc 3306 support via telco model (@SkylerBlumer)
- add enable support to ciscosmb (@deesel)
- add Waystream iBOS model
- add QTECH model (@moisseev)

### Changed

- more secret scrubbing in sonicos (@s-fu)
- openssh key scrubbing as secret in fortios (@agabellini)
- scrubs macsec key from Arista EOS (@krisamundson)
- rubocop dependency now ~> 0.80.0
- rugged dependency now ~> 0.28.0
- cumulus model no longer records transient data (@plett)

### Fixed

- use without-paging variant of print statement in routeros (@vkushnir)
- include the commands in the output in EdgeCOS model (@freddy36)
- update patterns for minor software version dependent differences in EdgeCOS model @freddy36)
- better login modalities for telnet in aos7 (@optimuscream)
- better virtual domain detection in fortios (@agabellini)
- allow any max length for username/password in GcomBNPS (@freddy36)
- relax prompt requirements in ciscosmb (@Atroskelis)
- fortios model strips uptime even without remove_secrets (@jplitza)
- HP ProCurve now accepts ">" as apart of the prompt (@magnuslarsen)
- fix IOS SNMP notification community hiding for informs and v3 (@moisseev)
- fixed issue where the regex-pattern for XOS-prompts used invalid syntax (@DarkCatapulter)
- set terminal width in EdgeCOS model (@moisseev)
- suppress errors for commands that are not supported on some devices in EdgeCOS model (@moisseev)
- revert including command names in the output of the EdgeCOS model (@moisseev)


lib/oxidized/config.rb
module Oxidized
  require 'asetus'
  class NoConfig < OxidizedError; end
  class InvalidConfig < OxidizedError; end
  class Config
    Root      = ENV['OXIDIZED_HOME'] || File.join(ENV['HOME'], '.config', 'oxidized')
    Crash     = File.join(ENV['OXIDIZED_LOGS'] || Root, 'crash')
    Log       = File.join(ENV['OXIDIZED_LOGS'] || Root, 'logs')
    InputDir  = File.join Directory, %w[lib oxidized input]
    OutputDir = File.join Directory, %w[lib oxidized output]
    ModelDir  = File.join Directory, %w[lib oxidized model]
    SourceDir = File.join Directory, %w[lib oxidized source]
    HookDir   = File.join Directory, %w[lib oxidized hook]
    Sleep     = 1

    def self.load(cmd_opts = {})
      asetus = Asetus.new(name: 'oxidized', load: false, key_to_s: true, usrdir: Oxidized::Config::Root)
      Oxidized.asetus = asetus

      asetus.default.username      = 'username'
      asetus.default.password      = 'password'
      asetus.default.model         = 'junos'
      asetus.default.resolve_dns   = true # if false, don't resolve DNS to IP
      asetus.default.interval      = 3600
      asetus.default.use_syslog    = false
      asetus.default.debug         = false
      asetus.default.threads       = 30
      asetus.default.use_max_threads = false
      asetus.default.timeout       = 20
      asetus.default.retries       = 3
      asetus.default.prompt        = /^([\w.@-]+[#>]\s?)$/
      asetus.default.rest          = '127.0.0.1:8888' # or false to disable
      asetus.default.next_adds_job = false            # if true, /next adds job, so device is fetched immmeiately
      asetus.default.vars          = {}               # could be 'enable'=>'enablePW'
      asetus.default.groups        = {}               # group level configuration
      asetus.default.group_map     = {}               # map aliases of groups to names
      asetus.default.models        = {}               # model level configuration
      asetus.default.pid           = File.join(Oxidized::Config::Root, 'pid')

      asetus.default.crash.directory = File.join(Oxidized::Config::Root, 'crashes')
      asetus.default.crash.hostnames = false

      asetus.default.stats.history_size = 10
      asetus.default.stats.redis_url    = false
      asetus.default.input.default      = 'ssh, telnet'
      asetus.default.input.debug        = false # or String for session log file
      asetus.default.input.ssh.secure   = false # complain about changed certs
      asetus.default.input.ftp.passive  = true  # ftp passive mode
      asetus.default.input.utf8_encoded = true  # configuration is utf8 encoded or ascii-8bit

      asetus.default.output.default = 'file'  # file, git
      asetus.default.source.default = 'csv'   # csv, sql

      asetus.default.model_map = {
        'juniper' => 'junos',
        'cisco'   => 'ios'
      }

      begin
        asetus.load # load system+user configs, merge to Config.cfg
      rescue StandardError => error
        raise InvalidConfig, "Error loading config: #{error.message}"
      end

      raise NoConfig, "edit #{File.join(Oxidized::Config::Root, 'config')}" if asetus.create

      # override if comand line flag given
      asetus.cfg.debug = cmd_opts[:debug] if cmd_opts[:debug]

      asetus
    end
  end

  class << self
    attr_accessor :mgr, :hooks
  end
end
